File: CodingTest/Coding_Test/programmers/2023 KAKAO BLIND RECRUITMENT/이모티콘 할인행사.java
import java.util.*;

class Solution {
    public static List<int[]> generateProducts(int[] arr,int r){
        List<int[]> products=new ArrayList<>();
        int[] product=new int[r];
        generateProduct(arr,r,product,0,products);

        return products;
    }

    public static void generateProduct(int[] arr,int r,int[] product, int index,List<int[]> products){
        if(r==index){
            products.add(product.clone());
            return;
        }
        for(int i=0;i<arr.length;i++){
            product[index]=arr[i];
            generateProduct(arr,r,product,index+1,products);
        }
    }

    public int[] solution(int[][] users, int[] emoticons) {
        int[] answer = {};
        int[] percentList={10,20,30,40};
        List<int[]> allCases=generateProducts(percentList,emoticons.length);
        int caseMaxServiceOwnerNum=0;
        int caseMaxTotal=0;
        for(int[] oneCase:allCases){
            int caseServiceOwner=0;
            int caseTotal=0;
            for(int[] user:users){
                int oneUserTotal=0;
                boolean userServiceOwned=false;
                for(int i=0;i<emoticons.length;i++){
                    if(oneCase[i]>=user[0]){
                        oneUserTotal+=(emoticons[i])*(100-oneCase[i])/100;
                    }
                    if(oneUserTotal>=user[1]){
                        caseServiceOwner+=1;
                        userServiceOwned=true;
                        break;
                    }
                }
                if(!userServiceOwned){
                    caseTotal+=oneUserTotal;
                }
            }
            if(caseServiceOwner>caseMaxServiceOwnerNum){
                caseMaxServiceOwnerNum=caseServiceOwner;
                caseMaxTotal=caseTotal;
            }else if(caseServiceOwner==caseMaxServiceOwnerNum&&caseTotal>caseMaxTotal){
                caseMaxTotal=caseTotal;
            }
        }

        answer=new int[]{caseMaxServiceOwnerNum,caseMaxTotal};
        return answer;
    }
}

File: CodingTest/Coding_Test/programmers/2023 KAKAO BLIND RECRUITMENT/이모티콘 할인행사.py
from itertools import product

def solution(users, emoticons):
    answer = []


    sales_percent=[10,20,30,40]

    cases=list(product(sales_percent,repeat=len(emoticons)))
    for case in cases:
        plus_service=0
        sales_total=0
        for user in users:
            total=0
            add_service=False
            for i in range(len(emoticons)):
                if case[i]>=user[0]:
                    total+=emoticons[i]*(100-case[i])//100
                if total>=user[1]:
                    plus_service+=1
                    add_service=True
                    break
            if not add_service:
                sales_total+=total
        answer.append([plus_service,sales_total])

    answer.sort(key=lambda x:(x[0],x[1]))
    # print(answer)
    return answer[-1]